vignettes/plugins.Rmd

```diff
@@ -0,0 +1,383 @@
+
+Taking this a step further, we can also specify a [schema](https://json-schema.org/) that this metadata will conform to
+
+```{r, echo = FALSE, results = "asis"}
+json_output(readLines(path_schema))
+```
+
+Now, when we register the plugin, we provide the path to this schema
+
+```{r}
+orderly2::orderly_plugin_register(
+  orderly2::orderly_plugin(db_config, db_read, db_prepare, path_schema),
+  "example.db")
+```
+
+Now, when the orderly metadata is saved (just before running the script part of a report) we will validate output that `db_prepare` generates against the schema, if `jsonvalidate` is installed (currently this requires our development version) and if the R option `outpack.schema_validate` is set to `TRUE` (e.g., by running `options(outpack.schema_validate = TRUE)`).
+
+If you need to control serialisation of the output from the "prepare" function, you should use `jsonlite::toJSON` within that function directly and orderly2 will interpolate it verbatim into the final JSON (the important part of this is that it ends up as a string with class `json`).
+
+# Arranging the plugin in a package
+
+Typically, a plugin should be in a package. Currently we only allow for a single plugin per package but later may relax this.
+
+All you need to do is make sure that the `.onLoad` function calls `orderly2:::orderly_register_plugin` - your package does not need to export any functions at all.
+
+```{r, export_to_package = TRUE}
+.onLoad <- function(...) {
+  orderly2::orderly_plugin_register(
+    orderly2::orderly_plugin(db_config, db_read, db_prepare, path_schema),
+    "example.db")
+}
+```
+
+The `name` argument to `orderly2:::orderly_register_plugin` must be the package name, as orderly2 will trigger loading the package based on this name in the configuration.
+
+```{r, include = FALSE}
+## This is a bit nasty, but saves keeping code in sync; collect up all the
+## chunks that have the option "export_to_package" together and concatenate
+## into a file
+code <- unname(Filter(
+  function(x) isTRUE(attr(x, "chunk_opts")$export_to_package),
+  knitr::knit_code$get()))
+path_pkg <- file.path(tempfile(), "example.db")
+fs::dir_create(path_pkg)
+
+writeLines(c(
+  "Package: example.db",
+  "Version: 0.0.1",
+  "License: CC0",
+  "Title: Orderly Database Example Plugin",
+  "Description: Simple example of an orderly plugin.",
+  "Authors@R: person('Orderly Authors', role = c('aut', 'cre'), email = 'email@example.com')",
+  "Imports: orderly2"),
+  file.path(path_pkg, "DESCRIPTION"))
+
+file.create(file.path(path_pkg, "NAMESPACE"))
+
+fs::dir_create(file.path(path_pkg, "R"))
+writeLines(
+  paste(vapply(code, paste, "", collapse = "\n"), collapse = "\n\n"),
+  file.path(path_pkg, "plugin.R"))
```

Comment:
Do you mean to write this into the `R` directory?

Reply:
yes - that must mean something later on in the vignette is busted too?

Reply:
The rest of the vignette looks ok atm, will need to update line 370 after moving `plugin.R`